Sources/StreamChat/Controllers/ChannelController/ChannelController_Tests.swift

```diff
@@ -3978,6 +3927,24 @@ class ChannelController_Tests: XCTestCase {
         // Assert controller is kept alive
         AssertAsync.staysTrue(weakController != nil)
     }
+
+    // MARK: Init registers active controller
+
+    func test_initRegistersActiveController() {
+        let client = ChatClient.mock
+        let channelQuery = ChannelQuery(cid: channelId)
+        let channelListQuery = ChannelListQuery(filter: .containMembers(userIds: [.unique]))
+
+        let controller = ChatChannelController(
+            channelQuery: channelQuery,
+            channelListQuery: channelListQuery,
+            client: client
+        )
+
+        XCTAssert(controller.client === client)
+        XCTAssert(client.activeChannelControllers.count == 1)
+        XCTAssert(client.activeChannelControllers.allObjects.first === controller)
+    }
```

Comment:
Both activeChannelControllers and activeChannelListControllers are weak HashTables, so there should not be any problem with that. If there would be a retain cycle, that would be because their implementation is faulty.

On top of that, I tried doing a basic test on the releasing part, and it is really complex to do because it is undetermined when NSHashTable will release its non referenced objects. Not even with autoreleasepools and dispatches I was able to simulate it. So as per unit tests, it is (nearly) impossible to get that.

Reply:
> Both activeChannelControllers and activeChannelListControllers are weak HashTables, so there should not be any problem with that. If there would be a retain cycle, that would be because their implementation is faulty.

But we don't wanna blindly rely on/beleive in the implementation we have, right? It can be or become faulty later and tests out insurance from the initial mistake/regression.

> On top of that, I tried doing a basic test on the releasing part, and it is really complex to do because it is undetermined when NSHashTable will release its non referenced objects. Not even with autoreleasepools and dispatches I was able to simulate it. So as per unit tests, it is (nearly) impossible to get that.

My first attempt involved `NSHashTable.weakObject` usage but I was not sure we won't try to access the component that has already been deallocated 🙂 This was the reason why I finally gave a preference to `Weak` wrapper.